### DISCUSSION TITLE
Glob matching shapes in and out of a container

### DISCUSSION THREAD

**Discussion by muteboy** [terrastruct/d2#1931] [Q&A] [answered]
I have a shape class called "question". I want to add a glob to make every shape with a name starting with "Question" use that class.
Some shapes are in containers, other are not. At the moment to match everything I have to do this:

```
Question*.class: question
**.Question*.class: question
```

Is there a way to do this in one line?

**Comment by mauroexe** [comment]
I came across a similar problem; sometimes you have to keep moving objects between containers and doing so you also have to redo all the relations... for giving you an example.

Using globs feels like hacking and it seems to be not efficient, I mean, compiling a file with globals takes more time.

Another approach is using "vars" like in this example:

```
vars {
  question:{
    #styles and object specifications
  }
}

anyObject:{@question}
```

You can use your vars inside any container, it feels faster and looks better than using globals, but I think it can be improved by having a "resources" section were you can define your objects for later use.

I describe my idea about "resources" here #1891 I really wish it can be implemented.

Hope it helps you.

**Reply by alixander** [reply]
> compiling a file with globals takes more time.

Can you give an example? (Maybe a D2 file I can run with and without globs). I haven't benchmarked it but the implementation shouldn't take noticeably more time.

**Reply by alixander** [reply]
And thanks for bumping that resources issue, will take another look to see if it should be prioritized now

**Reply by mauroexe** [reply]
For sure,  Alixander, I'll make an example this weekend :)

**Comment by alixander** [comment]
Oh interesting. `**.` should match root items too.

**Reply by muteboy** [reply]
Hi, just checking, was this resolved?

**Reply by alixander** [reply, marked as answer]
I realized that there actually is a way to do this (it's possible the additional functionality only came out after the original question you posted). You use `**` to glob every object, and filter by its label.

```d2
classes: {
  question: {
    label: ?
  }
}

**: {
  &label: question*
  class: question
}

question1

a: {
  question2
}
```

https://play.d2lang.com/?script=Ss5JLC5OLbZSqOZSUCgsTS0uyczPg_AUFHISk1JzrBTsuRQUarlqubi0tCAyalAJmHotLgUFsEEIIZByGNuQiysR1QIjrlouQAAAAP__&
